sklearn/base.py

```diff
@@ -349,7 +350,11 @@ def _check_n_features(self, X, reset):
                call to `partial_fit`. All other methods that validate `X`
                should set `reset=False`.
         """
-        n_features = X.shape[1]
+        try:
+            n_features = _num_features(X)
+        except TypeError:
+            # If the number of features is not defined skip this check
+            return
```

Comment:
Skip or raise an error?

Reply:
I wanted to make `_check_n_features` as lenient as possible, so metaestimators can call it with inputs and would not need to worry about this failing.

On that note, I think there is an argument for failing here when `reset=False` and `hasattr(self, "n_features_in_")`.

Reply:
I agree: if we can detect the number of feature at fit time, we will be strict at predict/transform time. But if we cannot, we are lenient.

Reply:
PR updated with my suggestion + tests